storage-proofs/src/crypto/feistel.rs

```diff
@@ -99,29 +102,49 @@ fn decode(index: u32, keys: &[u32], precomputed: FeistelPrecomputed) -> u32 {
     (left << half_bits) | right
 }
 
+const HALF_FEISTEL_BYTES: usize = mem::size_of::<Index>();
+const FEISTEL_BYTES: usize = 2 * HALF_FEISTEL_BYTES;
+
 // Round function of the Feistel network: `F(Ri, Ki)`. Joins the `right`
 // piece and the `key`, hashes it and returns the lower `u32` part of
 // the hash filtered trough the `right_mask`.
-fn feistel(right: u32, key: u32, right_mask: u32) -> u32 {
-    let mut data: [u8; 8] = [0; 8];
-    data[0] = (right >> 24) as u8;
-    data[1] = (right >> 16) as u8;
-    data[2] = (right >> 8) as u8;
-    data[3] = right as u8;
-
-    data[4] = (key >> 24) as u8;
-    data[5] = (key >> 16) as u8;
-    data[6] = (key >> 8) as u8;
-    data[7] = key as u8;
```

Comment:
I compared the assembly with an unrolled version, and it looks much better: https://godbolt.org/z/2rY9kd I would suggest using unrolled versions for all of these, and use two different unrolled methods if you want to make this compatible

Reply:
Fair enough, will do that.

Reply:
Someone once told me (cough) that iterators in Rust were so performant I should prefer them as cost-free abstractions. (Just giving you a hard time.)

Reply:
they are in many cases, which is why looked at the assembly :)

Reply:
You're the best — thanks for checking and showing your work. Godbolt is cool.